apps/bbsync/query.py

```diff
@@ -0,0 +1,184 @@
+            self._query["resolution"] = self.flaw.resolution
+
+    def generate_alias(self):
+        """
+        generate add or remove CVE alias query
+        conditionally based on the changes and create|update
+        """
+        if self.old_flaw is None:
+            if self.flaw.cve_id is not None:
+                # create query requires pure list
+                self._query["alias"] = [self.flaw.cve_id]
+
+        elif self.flaw.cve_id != self.old_flaw.cve_id:
+            self._query["alias"] = {}
+
+            if self.flaw.cve_id is not None:
+                self._query["alias"]["add"] = [self.flaw.cve_id]
+
+            if self.old_flaw.cve_id is not None:
+                self._query["alias"]["remove"] = [self.old_flaw.cve_id]
+
+    def generate_keywords(self):
+        """
+        generate keywords query based on creation|update
+        """
+        self._query["keywords"] = (
+            ["Security"] if self.old_flaw is None else {"add": ["Security"]}
+        )
+
+    def generate_flags(self):
+        """
+        generate query for Bugzilla flags
+        """
+        self._query["flags"] = []
+        # TODO needinfo and other flags
+        # hightouch | hightouch-lite | nist_cvss_validation | requires_doc_text
+
+    EMBARGOED_GROUPS = ["qe_staff", "security"]
+    DEADLINE_FORMAT = "%Y-%m-%d"
+
+    def generate_groups(self):
+        """
+        generate query for Bugzilla groups
+        which control the access to the flaw
+        plus eventually set deadline
+        """
+        # TODO groups handling is more complicated and involves
+        # product definitions processing but let us do it simple now
+
+        if self.flaw.embargoed:
+            if self.old_flaw is None:
+                self._query["groups"] = self.EMBARGOED_GROUPS
+
+            if self.flaw.unembargo_dt:
+                self._query["deadline"] = self.flaw.unembargo_dt.strftime(
+                    self.DEADLINE_FORMAT
+                )
+
+        # unembargo
+        elif self.old_flaw and self.old_flaw.embargoed:
+            # TODO we have no detailed information on previous groups
+            # so if there were some unusual we cannot guess
+            self._query["groups"] = {"remove": self.EMBARGOED_GROUPS}
+            self._query["deadline"] = ""
+
+        # TODO other case should be theoretically not needed to handle
+        # which is probably quite naive so let us test it extensively
+        #
+        # - embargoing of a public flaw is futile
+        #
+        # - groups should be set to embargoed or none
+        #   but there are definitely some corner cases
+        #
+        # - we need to store them on flaw fetch if we
+        #   want to do something more clever about them
+
+    def generate_cc(self):
+        """
+        generate query for CC list
+        """
+        # TODO now just empty CC list
+        self._query["cc"] = [] if self.old_flaw is None else {}
```

Comment:
This is some sort of a BZ magic that upon creation the `list` is expected and otherwise empty `dict`? :smile:

Reply:
That is rather my mistake :laughing: